[Docs] New Drivers and ORMs Section (#10878)
yugabyte/yugabyte-db: docs/content/latest/drivers-orms/python/sqlalchemy.md

---
title: SQLAlchemy ORM
linkTitle: SQLAlchemy ORM
description: SQLAlchemy ORM support for YugabyteDB
headcontent: SQLAlchemy ORM support for YugabyteDB
image: /images/section_icons/sample-data/s_s1-sampledata-3x.png
menu:
  latest:
    name: Python ORMs
    identifier: sqlalchemy-orm
    parent: python-drivers
    weight: 600
isTocNested: true
showAsideToc: true
---

> An Object Relational Mapping (ORM) tool is used by the developers to handle database access, it allows developers to map their object-oriented domain classes into the database tables. It simplifies the CRUD operations on your domain objects and easily allow the evolution of Domain objects to be applied to the Database tables.

Author: Done.

<ul class="nav nav-tabs-alt nav-tabs-yb">

  <li >
    <a href="/latest/drivers-orms/python/django/" class="nav-link">
      <i class="icon-java-bold" aria-hidden="true"></i>
      Django ORM
    </a>
  </li>

  <li >
    <a href="/latest/drivers-orms/python/sqlalchemy/" class="nav-link active">
      <i class="icon-postgres" aria-hidden="true"></i>
      SQLAlchemy ORM
    </a>
  </li>

</ul>

[SQLAlchemy](https://www.sqlalchemy.org/) is a popular ORM provider for Python applications, and is widely used by Python developers for database access. YugabyteDB provides full support for SQLAlchemy ORM.

### Add the SQLAlchemy ORM Dependency

To download and install SQLAlchemy to your project, use the following command.
```shell
  pip3 install sqlalchemy
  ```

  You can verify the installation as follows: 

  - Open the Python prompt by executing the following command:

    ```shell
    python3
    ```

  - From the Python prompt, execute the following commands to check the SQLAlchemy version:

    ```python prompt
    import sqlalchemy
    ```

    ```python prompt
    sqlalchemy.__version__
    ```

### Implementing ORM mapping for YugabyteDB

> To start with SQLAlchemy, in your project directory, create 4 Python files - `config.py`,`base.py`,`model.py` and `main.py`

Reviewer: ```suggestion
To start with SQLAlchemy, in your project directory, create 4 Python files - `config.py`,`base.py`,`model.py`, and `main.py`
```

> The `config.py` will contain the credentials to connect to your database. Copy the following sample code to your `config.py` file.

Reviewer: ```suggestion
`config.py` contains the credentials to connect to your database. Copy the following sample code to the `config.py` file.
```

```python
  db_user = 'yugabyte'
  db_password = 'yugabyte'
  database = 'yugabyte'
  db_host = 'localhost'
  db_port = 5433
```

> Next, declare a mapping. When using the ORM, the configuration process begins with describing the database tables you'll be using, and then defining the classes which will be mapped to those tables. In modern SQLAlchemy, these two tasks are usually performed together, using a system known as "Declarative Extensions". Classes mapped using the Declarative system are defined in terms of a base class which maintains a catalog of classes and tables relative to that base - this is known as the declarative base class.

Reviewer: ```suggestion
Next, declare a mapping. When using the ORM, the configuration process begins with describing the database tables you'll use, and then defining the classes which map to those tables. In modern SQLAlchemy, these two tasks are usually performed together, using a system known as "Declarative Extensions". Classes mapped using the Declarative system are defined in terms of a base class which maintains a catalog of classes and tables relative to that base - this is known as the declarative base class.
```

You create the Base class using the `declarative_base()` function. Add the following code to the `base.py` file.

```python
from sqlalchemy.ext.declarative import declarative_base
Base = declarative_base()
```

> Now that you have a “base”, you can define any number of mapped classes in terms of it. Start with a single table called `employees`, which will store records for the end-users using your application. A new class called `Employee` will be the class to which we map this table. Within the class, you define details about the table to which you'll be mapping, primarily the table name, and names and datatypes of the columns.

Reviewer: ```suggestion
Now that you have a “base”, you can define any number of mapped classes in terms of it. Start with a single table called `employees`, to store records for the end-users using your application. A new class called `Employee` maps to this table. In the class, you define details about the table to which you're mapping, primarily the table name, and names and datatypes of the columns.
```
